technical-reports/format/composite-types.md

```diff
@@ -60,155 +75,205 @@ Other examples might be:
+      }
+    ]
+  }
+}
+```
 
-### Appending custom type definitions
+Describes a gradient that is solid yellow for the first 2/3 and then fades to red:
 
-A tool might not understand certain user-defined type definitions in a token file and therefore can't do anything special with them. However, that tool could export or modify token files by adding their own type definitions.
+<div style="height: 2rem; background: linear-gradient(90deg, #ffff00 66.6%, #ff0000);"></div>
 
-For example, imagine Sketch wanted to export all layer styles from a Sketch document to a token file. The Sketch app could have a built-in type definition that corresponds to layer styles. Perhaps something like this:
+</aside>
 
-<aside class="example" title="Custom type definitions for Sketch">
+<aside class="example" title="Gradient token using references example">
 
 ```json
 {
-  "sketch layer style": {
-    "type": "typedef",
-    "value": {
-      "fill": {
-        "type": "color",
-        "required": false
+  "brand-primary": {
+    "type": "color",
+    "value": "#99ff66"
+  },
+
+  "position-end": {
+    "value": 1
+  },
+
+  "brand-in-the-middle": {
+    "type": "gradient",
+    "value": [
+      {
+        "color": "#000000",
+        "position": 0
       },
-      "border": {
-        "color": {
-          "type": "color",
-          "required": false
-        },
-        "width": {
-          "type": "dimension",
-          "required": false
-        }
+      {
+        "color": "{brand-primary}",
+        "position": 0.5
+      },
+      {
+        "color": "#000000",
+        "position": "{position-end}"
       }
-    }
+    ]
   }
 }
 ```
 
+Describes a color token called "brand-primary", which is referenced as the mid-point of a gradient is black at either end.
+
+<div style="height: 2rem; background: linear-gradient(90deg, #000000, #99ff66, #000000);"></div>
+
 </aside>
 
-It could then save out that type definition into the token file it exports. Other tools could then make use of that same type definition and do useful things with tokens of that type.
+## Typography
 
-### Custom configuration
+Represents a typographic style. The type property must be set to the string “typography”. The value must be an object with the following properties:
 
-Via configuration files or settings menus, tools could let users define custom behavior for types they have defined. For instance, in an export tool like Style Dictionary, users might be able to configure a custom transform or formatter for their composite types.
-Then, this example...
+- `fontName`: The typography's font. The value of this property must be a valid [font name](#font-name) or a reference to a font name token.
+- `fontSize`: The size of the typography. The value of this property must be a valid [dimension value](#dimension) or a reference to a dimension token.
+- `fontWeight`: The weight of the typography. The value of this property must be a valid JSON string or a reference to a string token.
```

Comment:
Should we restrict this to a pre-defined set of values (effectively, a sort of enum). E.g. `"bold"`, `"normal"`, `"light"`, etc. and/or numeric values like `100`, `200`, `300`, etc.?

Otherwise, any string is technically valid, so how should a tool interpret something like `"fontWeight": "hello"`?

If we do limit the values, I think it might make sense to define that as its own type (maybe "font-weight" or something like that)? It's conceivably something people might also define individual tokens for.

Reply:
@drwpow Thanks a lot for sharing your thoughts on this. The more feedback we get from the community the better it is tbh.

I agree with @drwpow regarding the `fontWeight` property. It's a tricky one because there's no clear convention.

In my opinion its value should be restricted to a set of string and numeric values.

String values:
- `ultrathin`
- `thin`
- `ultralight`
- `light`
- `regular`
- `book`
- `normal`
- `bold`
- `ultrabold`

I might have forgotten other frequently used values here though. The tricky part is that there's no strict convention between typefaces for these values.

Numeric values from `1` to `999`.

Reply:
Inspired by @ChucKN0risK & @drwpow's comments above, I propose the following:

* We add a new (non-composite) `fontWeight` type
* The `fontWeight` sub-value of the `typography` composite type must have values that are valid `fontWeight` token values or references to `fontWeight` tokens.

The new `fontWeight` type's values must be either a number value in the range `1` - `1000` (which seems to be the range [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Fonts/Variable_Fonts_Guide#weight) and [Android](https://developer.android.com/reference/kotlin/androidx/compose/ui/text/font/FontWeight) permit) _or_ one of the pre-defined strings listed in the table below. As per the table, the string names are just human-friendly aliases for certain number values, just like [how CSS does it](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight#common_weight_name_mapping), which in turn seems to be based on the [OpenType spec](https://docs.microsoft.com/en-us/typography/opentype/spec/os2#wtc):

| numeric value | string value aliases |
|-----------------|--------------------- |
| `100` | `thin`, `hairline` |
| `200` | `extraLight`, `ultraLight` |
| `300` | `light` |
| `400` | `normal`, `regular`, `book` | 
| `500` | `medium` |
| `600` | `semiBold`, `demiBold` |
| `700` | `bold` |
| `800` | `extraBold`, `ultraBold` |
| `900` | `black`, `heavy` |
| `950` | `extraBlack`, `ultraBlack` |

Sound good?

Reply:
I support this proposal @c1rrus

Reply:
Great! I've just pushed a commit that adds it and updates the `typography` type accordingly.

I went for a kebab case `font-weight` name rather than camel case as I used in my comment above in order to be consistent with the `cubic-bezier` type we already have. I also renamed the existing `font` type to `font-name` to make its purpose less ambiguous.

I wonder whether we should also change the sub-value names we've used in composite types to be kebab-case for the sake of consistency (or else, make everything camel case). What do you think?

Reply:
> I wonder whether we should also change the sub-value names we've used in composite types to be kebab-case for the sake of consistency (or else, make everything camel case). What do you think?
@c1rrus You're suggesting to rename `extraLight` as `extra-light`?
If so, I agree.

Reply:
@ChucKN0risK not quite. I already used kebab case for the values (e.g. `extra-light`). But the names of the typography type's sub-values are currently camel case:

```jsonc
{
  "example token": {
    "type": "typography",
    "value": {
      "fontWeight": "extra-light" // <-- this looks weird to me

      // ...
    }
  }
}
```
So, what I am proposing is that we make those kebab-case too: `"font-weight": "extra-light"`.

Furthermore, we should then adopt this convention across our entire format.

Reply:
+1 to `kebab-case` everywhere